### PR TITLE
Support SubstanceGroups and StereoGroups in JSON

### PR DESCRIPTION
**Pull request by greglandrum**
Modifications to the JSON parser/writer to support the `rdkitjson` format. This extends commonchem to support enhanced stereo and substance groups.

This also documents the changes and fixes a crashing bug in the JSON writer.

I didn't just add this to the RDKit representation part of the commonchem JSON because these two features are not RDKit specific and the representation could be easily used in other toolkits. These should be proposed as a commonchem extension at some point.

### CODE REVIEW

File: Code/GraphMol/MolInterchange/Writer.cpp [hunk 1]
```diff
@@ -230,7 +253,134 @@ void addBond(const Bond &bond, rj::Value &rjBond, rj::Document &doc,
...
+    } catch (const boost::bad_any_cast &) {
+      try {
+        auto val = obj.template getProp<double>(pN);
+        rjv = val;
+      } catch (const boost::bad_any_cast &) {
+        try {
+          auto val = obj.template getProp<std::string>(pN);
+          rjv.SetString(val.c_str(), val.size(), doc.GetAllocator());
+        } catch (const boost::bad_any_cast &) {
+          continue;
```
Comment (ricrogz) [inline Code/GraphMol/MolInterchange/Writer.cpp:273]: Is it worth throwing a warning message if we cannot cast the property, and end up skipping it?

File: Code/GraphMol/MolInterchange/Parser.cpp [hunk 2]
```diff
@@ -205,6 +207,143 @@ void readBond(RWMol *mol, const rj::Value &bondVal,
...
+  PRECONDITION(mol, "no mol");
+
+  for (const auto &sgVal : sgVals.GetArray()) {
+    if (!sgVal.HasMember("properties") ||
+        !sgVal["properties"].HasMember("TYPE")) {
+      throw FileParseException(
+          "Bad Format: substance group does not have TYPE property");
+    }
+
+    auto sgType = sgVal["properties"]["TYPE"].GetString();
```
Comment (ricrogz) [inline Code/GraphMol/MolInterchange/Parser.cpp:257]: Should we check whether the SGroup type is one of the known types, using `SubstanceGroupChecks::isValidType` ?

Also, after parsing the properties, we might want to check the subtype and "Connect" type, if these are present.

File: Code/GraphMol/MolInterchange/Parser.cpp [hunk 3]
```diff
@@ -205,6 +207,143 @@ void readBond(RWMol *mol, const rj::Value &bondVal,
...
+    SubstanceGroup sg(mol, sgType);
+
+    if (sgVal.HasMember("properties")) {
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/MolInterchange/Parser.cpp:260]: yep, that was copy pasta

File: Code/GraphMol/MolInterchange/Parser.cpp [hunk 4]
```diff
@@ -205,6 +207,143 @@ void readBond(RWMol *mol, const rj::Value &bondVal,
...
+      parseProperties(sg, sgVal["properties"]);
+    }
+
+    if (sgVal.HasMember("atoms")) {
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/MolInterchange/Parser.cpp:264]: I agree that checking in `addSubstanceGroup()` makes sense.

Ok if I do that in a separate PR?

[the remark this answers is not in the document]
Comment (ricrogz) [reply Code/GraphMol/MolInterchange/Parser.cpp:292]: Sure!

**Review by ricrogz — APPROVED** [review]
LGTM

**Pull request merged**